common/src/main/java/com/microsoft/identity/common/internal/commands/GenerateShrCommand.java

```diff
@@ -0,0 +1,111 @@
+//  furnished to do so, subject to the following conditions :
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.common.internal.commands;
+
+import androidx.annotation.NonNull;
+
+import com.microsoft.identity.common.exception.BaseException;
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.exception.UiRequiredException;
+import com.microsoft.identity.common.internal.commands.parameters.GenerateShrCommandParameters;
+import com.microsoft.identity.common.internal.controllers.BaseController;
+import com.microsoft.identity.common.internal.result.GenerateShrResult;
+
+import java.util.List;
+
+import static com.microsoft.identity.common.exception.ErrorStrings.NO_ACCOUNT_FOUND;
+
+/**
+ * Command class to perform generation of AT-less SHRs on behalf of a user.
+ */
+public class GenerateShrCommand extends BaseCommand<GenerateShrResult> {
+
+    private static final String TAG = GenerateShrCommand.class.getSimpleName();
+
+    /**
+     * Constructs a new GenerateShrCommand.
+     *
+     * @param parameters  The command's input parameters.
+     * @param controllers The controllers on which to run this command.
+     * @param callback    The command to notify once execution has completed.
+     * @param publicApiId The public API ID of this command.
+     */
+    public GenerateShrCommand(@NonNull final GenerateShrCommandParameters parameters,
+                              @NonNull final List<BaseController> controllers,
+                              @NonNull final CommandCallback<GenerateShrResult, BaseException> callback,
+                              @NonNull final String publicApiId) {
+        super(parameters, controllers, callback, publicApiId);
+    }
+
+    @Override
+    public GenerateShrResult execute() throws Exception {
+        final String methodName = ":execute";
+
+        GenerateShrResult result = null;
+        final GenerateShrCommandParameters parameters = (GenerateShrCommandParameters) getParameters();
+
+        // Iterate over our controllers, to service the request either locally or via the broker...
+        // if the local (embedded) cache contains tokens for the supplied user, we will sign using
+        // the embedded PoP keys. If no local user-state exists, the broker will be delegated to
+        // where the same check is performed.
```

Comment:
How are we certain that the `local` controller is the first one in the list here?

Reply:
This works the same way as the `SilentTokenCommand`: by calling `MsalControllerFactory#getAllControllers` where the `local` controller is always the first one tried. See:
https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msal/src/main/java/com/microsoft/identity/client/internal/controllers/MSALControllerFactory.java#L100

Reply:
Marking resolved; feel free to unresolve if you have more feedback on this